collect_app/src/test/java/org/odk/collect/android/utilities/MediaUtilsTest.kt

```diff
@@ -0,0 +1,80 @@
+package org.odk.collect.android.utilities
+
+import android.content.Context
+import android.content.Intent
+import android.net.Uri
+import android.widget.Toast
+import androidx.test.core.app.ApplicationProvider
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import org.hamcrest.CoreMatchers.`is`
+import org.hamcrest.MatcherAssert.assertThat
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.kotlin.argumentCaptor
+import org.mockito.kotlin.mock
+import org.mockito.kotlin.verify
+import org.mockito.kotlin.whenever
+import org.odk.collect.android.BuildConfig
+import org.odk.collect.androidshared.system.IntentLauncher
+import org.robolectric.shadows.ShadowToast
+import java.io.File
+
+@RunWith(AndroidJUnit4::class)
+class MediaUtilsTest {
+
+    private val context = ApplicationProvider.getApplicationContext<Context>()
+    private lateinit var intentLauncher: IntentLauncher
+    private lateinit var contentUriProvider: ContentUriProvider
+    private lateinit var mediaUtils: MediaUtils
+
+    @Before
+    fun setup() {
+        intentLauncher = mock()
+        contentUriProvider = mock()
+        this.mediaUtils = MediaUtils(intentLauncher, contentUriProvider)
+    }
+
+    @Test
+    fun `When file that we try to open does not exist a toast should be displayed`() {
+        val file = File("file://image.png")
+
+        mediaUtils.openFile(context, file, "image/*")
+
+        assertThat(ShadowToast.getTextOfLatestToast(), `is`("File: file:/image.png is missing."))
+        assertThat(ShadowToast.getLatestToast().duration, `is`(Toast.LENGTH_LONG))
+    }
+
+    @Test
+    fun `When open file, launch on intentLauncher should be called with proper parameters`() {
+        val file = File.createTempFile("image", ".png")
+        whenever(
+            contentUriProvider.getUriForFile(
+                context,
+                BuildConfig.APPLICATION_ID + ".provider",
+                file
+            )
+        ).thenReturn(
+            Uri.parse("content://org.odk.collect.android.provider/external_files/Android/data/org.odk.collect.android/files/image.png")
+        )
+
+        val contextCaptor = argumentCaptor<Context>()
+        val intentCaptor = argumentCaptor<Intent>()
+        val runnableCaptor = argumentCaptor<() -> Unit>()
+
+        mediaUtils.openFile(context, file, "image/*")
+
+        verify(intentLauncher).launch(
```

Comment:
An alternative to this that might be more readable (that using agument captors) is to build an `Intent` for the expectation. I think `Intent.equals` works pretty well if I recall correctly.

Reply:
The problem is that we have a function as a parameter in `launch`. I'm not interested in verifying it here but also I can't use `any()` since mixing `any()` and real object when you verify parameters is not possible. If I'm missing something about testing cases like this one please let me know.

Reply:
Oh you can do `any()` and object assertions together, but you have to use `eq()` for the object assertions rather than just passing the object as the parameter. The error message you get back from mixing `any()` and plain objects is really not helpful!

As an aside, if you find you need to use an argument captor, it can often be a sign that it'd be easier to make the mock manually by extending the class than using mockito.

Reply:
Ok but intent does not override equals so I can't use eq() for that since it would always check referential equality.

Reply:
Ach, my bad I thought `Intent` had a nice `equals()`. Never mind!